### PR TITLE
Add functions to see if a provider is available for use.

### PR DESCRIPTION
**Pull request by levitte**
Public function OSSL_PROVIDER_available() takes a library context and
a provider name, and returns 1 if it's available for use, i.e. if it's
possible to fetch implementations from it, otherwise 0.

Internal function ossl_provider_activated() returns 1 if the given
OSSL_PROVIDER is activated, otherwise 0.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Note that this new functionality is tested in `test/evp_test.c`

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod]: This is missing from the NAME section (hence the travis failure)

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod]: Can OPENSSL_CTX be NULL?

Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod]: What happens if no providers have been loaded yet, but this is a fallback provider?

File: test/evp_test.c
Comment (mattcaswell) [inline test/evp_test.c]: If the legacy provider has been built then we should always run these tests with the legacy provider available.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: > Can OPENSSL_CTX be NULL?

Does this really need to be repeated *everywhere*?

> What happens if no providers have been loaded yet, but this is a fallback provider?

Will rephrase

File: test/evp_test.c
Comment (mattcaswell) [reply test/evp_test.c]: In fact we may want to run this twice. Once with default and legacy available, and once with fips available (but no default or legacy).

Comment (levitte) [reply test/evp_test.c]: Yes.  My thought is that this would be controlled with a config file.

Comment (mattcaswell) [reply test/evp_test.c]: > Yes. My thought is that this would be controlled with a config file.

If so then that needs to be in scope of this PR. Otherwise we are losing a chunk of the tests with this PR.

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod]: > Does this really need to be repeated everywhere?

Yes, I think so. Or if not then a reference to where it is stated should be given everywhere. These are man pages so we have to assume that people dip in and out of them without the context of having read other pages.

File: test/evp_test.c
Comment (levitte) [reply test/evp_test.c]: > If so then that needs to be in scope of this PR.

Autoloading of config files is part of another PR.  However, I can certainly add the necessary lines in `test/recipes/30-test_evp.t` to do this, even though currently useless (because autoloading isn't in place).  That will start working as soon as the autoloading PR is merged.
Is that what you're asking for?

Comment (mattcaswell) [reply test/evp_test.c]: If we are relying on another PR for this to work then this one shouldn't be merged until that one is. Please add the necessary lines.

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: I prefer a link to OPENSSL_CTX(3)

Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: Done

Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod]: Hmm. Just a link in the "See Also" section doesn't seem sufficient.

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: Fixed

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: Why not?  If someone wants to know more about OPENSSL_CTX, then they should read the manual page for OPENSSL_CTX, doesn't that make sense?

File: test/evp_test.c
Comment (levitte) [reply test/evp_test.c]: Will do

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod]: If I was reading the man page for a function I would expect to see some discussion of all of the parameters for that function even if it is just to say "...see L<OPENSSL_CTX(3)> for further details of the B\<ctx\> parameter". The fact that a separate OPENSSL_CTX man page exists may not be immediately obvious, and having its reference buried away in the SEE ALSO at the very bottom of the page is likely to be missed.

File: CHANGES
Comment (t8m) [inline CHANGES]: new function

Comment (levitte) [reply CHANGES]: Fixed

File: test/evp_test.c
Comment (levitte) [reply test/evp_test.c]: Done

File: test/recipes/30-test_evp.t
Comment (mattcaswell) [inline test/recipes/30-test_evp.t]: Perhaps for this test it is sufficient just to run `default-and-legacy.cnf` and `fips.cnf`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/recipes/30-test_evp.t
Comment (levitte) [reply test/recipes/30-test_evp.t]: Sure...

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: I added a general line in the general description.  I'm starting to think that's what we should do, generally, rather than have it attached to each function (which is its own kind of buried)

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod]: I think this could do with some expansion around the behaviour with fallbacks. Consider for example this code:

````
int defavl = OSSL_PROVIDER_available(NULL, "default");
OSSL_PROVIDER *legacy = OSSL_PROVIDER_load(NULL, "legacy");
EVP_MD *md;

if (defavl) {
    md = EVP_MD_fetch(NULL, "SHA256", NULL);
} else {
    md = EVP_MD_fetch(NULL, "whirlpool", NULL);
}

    /* Do something with md */
````

My understanding from the code in this PR is that `OSSL_PROVIDER_available()` will return true if the provider is marked as a fallback, even if it hasn't actually been activated. So in the example above if no providers have been explicitly loaded then the `OSSL_PROVIDER_available()` call will return true, even though the subsequent load of the legacy provider means it is no longer available. This could be quite confusing.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c
Comment (mattcaswell) [inline crypto/provider_core.c]: So what happens if other providers are loaded but `prov` is a fallback provider that hasn't been initialized? Doesn't that mean `ossl_provider_available` will return true even though the provider isn't currently loaded and won't ever be with an explicit call to `OSSL_PROVIDER_load`?

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: And otherwise, if asked early, the default will look unavailable...

Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: The solution may be for this function to trigger the fallback process. Then it will be able to give a deterministic answer.

Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod]: Yes - that makes a lot of sense.

File: test/evp_test.c
Comment (slontis) [inline test/evp_test.c]: Should we skip or run and expect failure?

Comment (levitte) [reply test/evp_test.c]: Skip, like the next few lines indicate... Or do I misunderstand the question?

Comment (slontis) [reply test/evp_test.c]: I am asking if instead of doing the skipping... It could run and expect to fail..

We at least need some negative testing somewhere, that tests that algorithms are not available when they shouldnt be...

Comment (levitte) [reply test/evp_test.c]: Hmmm, so something conditional. That would be a conditional "Reason" the... Trouble is that evp_test isn't at all designed for such things, at least now. It's an idea, sure, but unfitting for this PR.

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: Okie, done.  Some refactoring was needed to get this working.

File: crypto/provider_core.c
Comment (mattcaswell) [inline crypto/provider_core.c]: Other places where we read `store->use_fallbacks` we are protected by the store lock. However if this function is called from `ossl_provider_available` then that isn't the case.

File: doc/internal/man3/ossl_provider_new.pod
Comment (mattcaswell) [inline doc/internal/man3/ossl_provider_new.pod]: This could do with tweaking - if fallbacks are not yet activated it will activate them (if appropriate), and *then* check if the provider is activated.

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod]: Should we state here about the behaviour wrt fallbacks?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c
Comment (levitte) [reply crypto/provider_core.c]: Fixed

File: doc/man3/OSSL_PROVIDER.pod
Comment (levitte) [reply doc/man3/OSSL_PROVIDER.pod]: No, that's internal

File: doc/internal/man3/ossl_provider_new.pod
Comment (levitte) [reply doc/internal/man3/ossl_provider_new.pod]: Done

**Comment by levitte** [comment]
Rebased

File: crypto/provider_core.c
Comment (mattcaswell) [reply crypto/provider_core.c]: Errr....is this fixed???

Comment (levitte) [reply crypto/provider_core.c]: Yes.  All calls to this function are guarded with a lock at this point.

File: doc/internal/man3/ossl_provider_new.pod
Comment (mattcaswell) [inline doc/internal/man3/ossl_provider_new.pod]: Typo: s/of/if/

**Review by mattcaswell — APPROVED** [review]
Approved assuming typo is fixed.

**Comment by levitte** [comment]
Thanks for the review!  Merged

36f5ec55e69716024f70df53074a2871e091a3e1 Add functions to see if a provider is available for use.
3b5d61f4721f91b5f31a8d3b935f9b3cf4c27644 test/evp_test.c: modify to use OSSL_PROVIDER_available()
e2f72313ccd168eb571b8a4c7cfaf0bf46bdcf9f test/recipes/30-test_evp.t: Modify to test with different providers

**Pull request closed without merging**
